applied-crypto-examples/src/encrypted_zksnark.rs

```diff
@@ -0,0 +1,254 @@
+//! An example of naive implementations of zksnarks for tutorial purposes
+use curve25519_dalek::{
+    constants::RISTRETTO_BASEPOINT_POINT, ristretto::RistrettoPoint, scalar::Scalar,
+};
+const G: RistrettoPoint = RISTRETTO_BASEPOINT_POINT;
+
+/// Single root of a polynomial
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Root {
+    // a in ax+b
+    a: Scalar,
+    // b in ax+b
+    b: Scalar,
+}
+
+impl Root {
+    /// Create a new root from i64 values
+    pub fn new(a: i64, b: i64) -> Result<Self, Error> {
+        if b % a == 0 {
+            let mut a_prime = Scalar::from(a.unsigned_abs());
+            let mut b_prime = Scalar::from(b.unsigned_abs());
+            if a < 0 {
+                a_prime = -a_prime;
+            }
+            if b < 0 {
+                b_prime = -b_prime;
+            }
+            return Ok(Self {
+                a: a_prime,
+                b: b_prime,
+            });
+        }
+        Err(Error::OutsideIntegerField(a, b))
+    }
+
+    /// Evaluate the root at a given scalar
+    pub fn eval(&self, x: &Scalar) -> Scalar {
+        self.a * x + self.b
+    }
+}
+
+/// Object providing an encrypted list of scalars encrypted by multiplying them by an
+/// elliptic curve (making them discrete log hard to calculate) and an evaluation of
+/// the public roots of the target polynomial that is able to verify prover responses.
+pub struct EncryptedChallenge {
+    /// List of Ristretto curve points created by multiplying the secret scalar by the
+    /// Ristretto basepoint
+    pub encrypted_powers: Vec<RistrettoPoint>,
+    // Public roots of the target polynomial evaluate at the secret scalar
+    ts: Scalar,
+}
```

Comment:
Potentially we should have a scalar generation function attached this?

Reply:
Possibly yes, especially if we're looking to create non-interactivity.